WeatherApp/WeatherApp/Clients/NetworkClient.swift

```diff
@@ -0,0 +1,56 @@
+import Foundation
+
+enum ClientError: Error {
+
+    case badURL
+    case decodingError(Error)
+    case networkError(Error)
+    case httpError(Int)
+    case noData
+    case unknown
+
+}
+
+protocol BaseApiClientProtocol {
+
+    func get<T: Decodable>(endpoint: Endpoint, completion: @escaping (Result<T, ClientError>) -> Void)
+
+}
+
+class NetworkClient: BaseApiClientProtocol {
+
+    func get<T: Decodable>(endpoint: Endpoint, completion: @escaping (Result<T, ClientError>) -> Void) {
+        guard let request = endpoint.buildRequest() else {
+            completion(.failure(.badURL))
+            return
+        }
+
+        let task = URLSession.shared.dataTask(with: request) { data, response, error in
+            if let error = error {
+                completion(.failure(.networkError(error)))
+                return
+            }
+
+            guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
+                completion(.failure(.httpError((response as? HTTPURLResponse)?.statusCode ?? -1)))
```

Comment:
`(response as? HTTPURLResponse)?.statusCode ?? -1` extract this and use as a variable to make it easier to read